### DISCUSSION TITLE
Support Sonarr API v4

### DISCUSSION THREAD

**Discussion by kaouk** [gethomepage/homepage#2607] [Feature Requests] [answered]
### Description

Since Sonarr has upgraded to v4 the API doesn't seem to work correctly with the embedded service widget

### Steps to reproduce

Setup services.yaml with the sonarr widget using API

### homepage version

v0.8.4

### Installation method

Docker

### Configuration

```yaml
- Sonarr:
          icon: sonarr.png
          href: https://>URL<
          server: >SERVER<
          container: sonarr
          widget:
              type: sonarr
              url: https://>URL<
              key: >KEY<
```

### Troubleshooting

Logs from homepage:

API Error: Invalid data
URL: https://>URL<//api/v3/wanted/missing?apikey=***
Response Data:
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta name="mobile-web-app-capable" content="yes" />
    <meta name="apple-mobile-web-app-capable" content="yes" />

    
    <meta name="theme-color" content="#3a3f51" />
    
    <meta name="msapplication-navbutton-color" content="#3a3f51" />

    <meta name="description" content="Sonarr" />

    <link
      rel="apple-touch-icon"
      sizes="180x180"
      href="/Content/Images/Icons/apple-touch-icon.png?h=d+iS4xWxn2A2bjUwrERmSg"
    />
    <link
      rel="icon"
      type="image/png"
      sizes="32x32"
      href="/Content/Images/Icons/favicon-32x32.png?h=s64FHSxrh1sgZCiBBIVikQ"
    />
    <link
      rel="icon"
      type="image/png"
      sizes="16x16"
      href="/Content/Images/Icons/favicon-16x16.png?h=wksVbPqhKGB2B5P0O0h8IQ"
    />
    <link rel="manifest" href="/Content/Images/Icons/manifest.json?h=/5qJyrGdXNWB0rzU5fobuA" crossorigin="use-credentials" />
    <link
      rel="mask-icon"
      href="/Content/Images/Icons/safari-pinned-tab.svg?h=LLFuxx74TMIn3Vx2ai6a1A"
      color="#00ccff"
    />
    <link
      rel="shortcut icon"
      type="image/ico"
      href="/favicon.ico?h=ZX67/60OXHGhAQ+8myebHg"
      data-no-hash
    />
    <meta
      name="msapplication-config"
      content="/Content/Images/Icons/browserconfig.xml"
    />

    <link rel="stylesheet" type="text/css" href="/Content/styles.css?h=VAwjWra8YBWH8JiyE7JFDA" />
    <link rel="stylesheet" type="text/css" href="/Content/Fonts/fonts.css?h=mBv1x1/yHnvphLclzeQIBA" />

    <title>Login - Sonarr</title>

    <style>
      body {
        background-color: #f5f7fa;
        color: #656565;
        font-family: "Roboto", "open sans", "Helvetica Neue", Helvetica, Arial,
          sans-serif;
      }

      .center {
        display: flex;
        justify-content: center;
      }

      .content {
        flex: 0 0 325px;
      }

      .logo {
        width: 32px;
        height: 32px;
      }

      .panel {
        margin-top: 50px;
        border-radius: 4px;
      }

      .panel-header {
        display: flex;
        justify-content: center;
        padding: 10px;
        border-top-left-radius: 4px;
        border-top-right-radius: 4px;
        background-color: #3a3f51;
      }

      .panel-body {
        padding: 20px;
        border-bottom-right-radius: 4px;
        border-bottom-left-radius: 4px;
        background-color: #fff;
      }

      .sign-in {
        padding: 10px 0;
        text-align: center;
      }

      .form-group {
        margin: 20px 0;
      }

      .form-input {
        box-sizing: border-box;
        padding: 6px 16px;
        width: 100%;
        height: 35px;
        border: 1px solid #dde6e9;
        border-radius: 4px;
        box-shadow: inset 0 1px 1px rgba(0, 0, 0, 0.075);
      }

      .form-input:focus {
        outline: 0;
        border-color: #66afe9;
        box-shadow: inset 0 1px 1px rgba(0, 0, 0, 0.075),
          0 0 8px rgba(102, 175, 233, 0.6);
      }

      .button {
        overflow: hidden;
        margin-top: 20px;
        padding: 10px 0;
        width: 100%;
        border: 1px solid;
        border-color: #5899eb;
        border-radius: 4px;
        background-color: #5d9cec;
        color: #fff;
        vertical-align: middle;
        text-align: center;
        white-space: nowrap;
        line-height: normal;
      }

      .button:hover {
        border-color: #3483e7;
        background-color: #4b91ea;
        color: #fff;
        text-decoration: none;
      }

      .copy {
        margin-top: 15px;
        text-align: center;
        font-size: 14px;
      }

      .remember-me-container {
        display: flex;
        justify-content: space-between;
        line-height: 25px;
      }

      .remember-me {
        font-size: 14px;
      }

      .forgot-password {
        margin-left: auto;
        color: #909fa7;
        text-decoration: none;
        font-size: 13px;
      }

      .forgot-password:focus,
      .forgot-password:hover {
        color: #748690;
        text-decoration: underline;
      }

      .forgot-password:visited {
        color: #748690;
      }

      .login-failed {
        margin-top: 20px;
        color: #f05050;
        font-size: 14px;
      }

      .hidden {
        display: none;
      }

      @media only screen and (min-device-width: 375px) and (max-device-width: 812px) {
        .form-input {
          font-size: 16px;
        }
      }
    </style>
  </head>

  <body>
    <div class="center">
      <div class="content">
        <div class="panel">
          <div class="panel-header">
            <img src="/Content/Images/logo.svg?h=lzId6hpOgICvOf7RS90NNQ" alt="Image" class="logo" />
          </div>

          <div class="panel-body">
            <div class="sign-in">
              SIGN IN TO CONTINUE
            </div>

            <form
              role="form"
              data-parsley-validate=""
              novalidate=""
              class="mb-lg"
              method="POST"
            >
              <div class="form-group">
                <input
                  type="email"
                  name="username"
                  class="form-input"
                  placeholder="Username"
                  autocomplete="off"
                  pattern=".{1,}"
                  required
                  title="User name is required"
                  autoFocus="true"
                  autoCapitalize="false"
                />
              </div>

              <div class="form-group">
                <input
                  type="password"
                  name="password"
                  class="form-input"
                  placeholder="Password"
                  required
                />
              </div>

              <div class="remember-me-container">
                <span class="remember-me">
                  <input
                    type="checkbox"
                    name="rememberMe"
                    id="rememberMe"
                    checked="checked"
                  />
                  <label for="rememberMe">Remember Me</label>
                </span>

                <a
                  href="https://wiki.servarr.com/sonarr/faq#help-i-have-locked-myself-out"
                  class="forgot-password"
                  >Forgot your password?</a
                >
              </div>
              <button type="submit" class="button">Login</button>

              <div id="login-failed" class="login-failed hidden">
                Incorrect Username or Password
              </div>
            </form>
          </div>
        </div>

        <div id="copy" class="copy hidden">
          <span>&copy;</span>
          <span id="year"></span>
          <span>-</span>
          <span>Sonarr</span>
        </div>
      </div>
    </div>
  </body>

  <script type="text/javascript">
    var yearSpan = document.getElementById("year");
    yearSpan.innerHTML = "2010-" + new Date().getFullYear();

    var copyDiv = document.getElementById("copy");
    copyDiv.classList.remove("hidden");

    if (window.location.search.indexOf("loginFailed=true") > -1) {
      var loginFailedDiv = document.getElementById("login-failed");

      loginFailedDiv.classList.remove("hidden");
    }
  </script>
</html>

### Before submitting, I have made sure to

- [X] Check [the documentation](https://gethomepage.dev/)
- [X] Follow [the troubleshooting guide](https://gethomepage.dev/latest/more/troubleshooting/) (please include output above if applicable).
- [X] Search [existing issues](https://github.com/gethomepage/homepage/search?q=&type=issues) and [discussions](https://github.com/gethomepage/homepage/search?q=&type=discussions).

**Comment by shamoon** [comment]
I would be a bit surprised if they broke things like this...

https://sonarr.tv/docs/api/:

> Sonarr API docs - The v3 API docs apply to both v3 and v4 versions of Sonarr. Some functionality may only be available in v4 of the Sonarr application.

**Reply by shamoon** [reply, marked as answer]
Yea, they didnt

<img width="392" alt="Screenshot 2024-01-10 at 3 44 13 PM" src="https://github.com/gethomepage/homepage/assets/4887959/c9731806-c49e-43bc-9e93-1d6cdfbbf0e9">
<img width="717" alt="Screenshot 2024-01-10 at 3 45 41 PM" src="https://github.com/gethomepage/homepage/assets/4887959/24647343-1fbe-4472-ac82-c089f8534096">


Please follow the troubleshooting guide as instructed in the "bug report"

**Comment by Schabernack** [comment]
I ran into the same breaking issue going from sonarr 3->4.   I fixed it by updating the sonarr url in the  homepage services.yaml widget definition. for some reason, it no longer likes trailing slashes

```
# bad
          url: http://192.168.2.42:8989/ 
# good
          url: http://192.168.2.42:8989
```

**Reply by shamoon** [reply]
https://gethomepage.dev/latest/more/troubleshooting/#service-widget-errors

See point 1

**Reply by kaouk** [reply]
Doh, that fixed it! Thanks :)

**Comment by kaouk** [comment]
> I ran into the same breaking issue going from sonarr 3->4. I fixed it by updating the sonarr url in the homepage services.yaml widget definition. for some reason, it no longer likes trailing slashes
> 
> ```
> # bad
>           url: http://192.168.2.42:8989/ 
> # good
>           url: http://192.168.2.42:8989
> ```

**Comment by kaouk** [comment]
Was an issue with trailing slash on the URL as noted by another user and seen in the troubleshooting guidance.

**Comment by jjscaria** [comment]
Apologies for bumping a closed discussion. I am having the same kind of issue and its not trailing slashes. Its making a call to API v3

Homepage version: 0.8.4

Sonarr version: 4.0.0.748

```
    - Shows:
        href: http://192.168.1.101:8989
        ping: http://192.168.1.101:8989
        icon: sonarr.png
        widget:
          type: sonarr
          url: http://192.168.1.101:8989
          key: apikeyapikey
```


![sonarr](https://github.com/gethomepage/homepage/assets/9114566/93a72b70-b108-4842-a509-734792e65d7a)
